tensorflow/python/kernel_tests/map_stage_op_test.py

```diff
@@ -0,0 +1,244 @@
+      stager = data_flow_ops.MapStagingArea(
+          [dtypes.float32, dtypes.float32],
+          shapes=[[], [128, 128]],
+          names=['x', 'v'])
+      stage = stager.put(pi,{'x': x, 'v': v})
+      size = stager.size()
+      clear = stager.clear()
+
+    with self.test_session(use_gpu=True) as sess:
+      sess.run(stage, feed_dict={x: -1, pi: 3})
+      self.assertEqual(sess.run(size), 1)
+      sess.run(stage, feed_dict={x: -1, pi: 1})
+      self.assertEqual(sess.run(size), 2)
+      sess.run(clear)
+      self.assertEqual(sess.run(size), 0)
+
+
+  def testCapacity(self):
+    capacity = 3
+
+    with ops.device('/cpu:0'):
+      x = array_ops.placeholder(dtypes.int32, name='x')
+      pi = array_ops.placeholder(dtypes.int64, name='pi')
+      gi = array_ops.placeholder(dtypes.int64, name='gi')
+    with ops.device(test.gpu_device_name()):
+      stager = data_flow_ops.MapStagingArea([dtypes.int32, ],
+        capacity=capacity, shapes=[[]])
+      stage = stager.put(pi,[x])
+      get = stager.get()
+      size = stager.size()
+
+    from six.moves import queue as Queue
+    import threading
+
+    queue = Queue.Queue()
+    n = 5
+    missed = 0
+
+    with self.test_session(use_gpu=True) as sess:
+      # Stage data in a separate thread which will block
+      # when it hits the staging area's capacity and thus
+      # not fill the queue with n tokens
+      def thread_run():
+        for i in range(n):
+          sess.run(stage, feed_dict={x: i, pi: i})
+          queue.put(0)
+
+      t = threading.Thread(target=thread_run)
+      t.start()
+
+      # Get tokens from the queue, making notes of when we timeout
+      for i in range(n):
+        try:
+          queue.get(timeout=0.05)
+        except Queue.Empty:
+          missed += 1
+
+      # We timed out n - capacity times waiting for queue puts
+      self.assertTrue(missed == n - capacity)
+
+      # Clear the staging area out a bit
+      for i in range(n - capacity):
+        sess.run(get)
+
+      # This should now succeed
+      t.join()
+
+      self.assertTrue(sess.run(size) == capacity)
+
+      # Clear out the staging area completely
+      for i in range(capacity):
+        sess.run(get)
+
+  def testOrdering(self):
+    import six
+    import random
+
+    with ops.device('/cpu:0'):
+      x = array_ops.placeholder(dtypes.int32, name='x')
+      pi = array_ops.placeholder(dtypes.int64, name='pi')
+      gi = array_ops.placeholder(dtypes.int64, name='gi')
+    with ops.device(test.gpu_device_name()):
+      stager = data_flow_ops.MapStagingArea([dtypes.int32, ],
+        shapes=[[]], ordered=True)
+      stage = stager.put(pi,[x])
+      get = stager.get()
+      size = stager.size()
+
+    n = 10
+
+    with self.test_session(use_gpu=True) as sess:
+      # Keys 0..n-1
+      keys = list(six.moves.range(n))
+
+      # Shuffle them for random insert
+      shuffle_keys = [i for i in keys]
+      random.shuffle(shuffle_keys)
```

Comment:
I think that this still has a small but finite probability of ending up with an unsorted list.  The shuffle could undo the reverse.  Just reverse.  Then you guarantee that the list is not sorted.